kong/dns/client.lua

```diff
@@ -0,0 +1,691 @@
+                                                 tries)
+  -- check for runtime errors in the callback
+  if err and err:sub(1, 8) == "callback" then
+    log(ALERT, PREFIX, err)
+  end
+
+  local hit_str = hit_level and HIT_LEVEL_TO_NAME[hit_level] or "fail"
+  stats_increment(self.stats, key, hit_str)
+
+  log(DEBUG, PREFIX, "cache lookup ", key, " ans:", answers and #answers or "-",
+                     " hlv:", hit_str)
+
+  if has_timing then
+    req_dyn_hook_run_hook("timing", "dns:cache_lookup",
+                          (hit_level and hit_level < HIT_L3))
+  end
+
+  if answers and answers.errcode then
+    err = ("dns %s error: %s %s"):format(
+            answers.errcode < CACHE_ONLY_ERROR_CODE and "server" or "client",
+            answers.errcode, answers.errstr)
+    return nil, err, tries
+  end
+
+  return answers, err, tries
+end
+
+
+function _M:resolve(name, qtype, cache_only, tries)
+  return resolve_all(self, name, qtype, cache_only, tries,
+                     ngx.ctx and ngx.ctx.has_timing)
+end
+
+
+function _M:resolve_address(name, port, cache_only, tries)
+  local has_timing = ngx.ctx and ngx.ctx.has_timing
+
+  local answers, err, tries = resolve_all(self, name, nil, cache_only, tries,
+                                          has_timing)
+
+  if answers and answers[1] and answers[1].type == TYPE_SRV then
+    local answer = get_next_weighted_round_robin_answer(answers)
+    port = answer.port ~= 0 and answer.port or port
+    answers, err, tries = resolve_all(self, answer.target, TYPE_A_AAAA,
+                                      cache_only, tries, has_timing)
+  end
+
+  if not answers then
+    return nil, err, tries
+  end
+
+  return get_next_round_robin_answer(answers).address, port, tries
+end
+
+
+-- compatible with original DNS client library
+-- These APIs will be deprecated if fully replacing the original one.
+local dns_client
+
+function _M.init(opts)
+  log(DEBUG, PREFIX, "(re)configuring dns client")
+
+  if opts then
+    opts.valid_ttl = opts.valid_ttl or opts.validTtl
+    opts.error_ttl = opts.error_ttl or opts.badTtl
+    opts.stale_ttl = opts.stale_ttl or opts.staleTtl
+    opts.cache_size = opts.cache_size or opts.cacheSize
+  end
+
+  local client, err = _M.new(opts)
+  if not client then
+    return nil, err
+  end
+
+  dns_client = client
+  return true
+end
+
+
+-- New and old libraries have the same function name.
+_M._resolve = _M.resolve
+
+function _M.resolve(name, r_opts, cache_only, tries)
+  return dns_client:_resolve(name, r_opts and r_opts.qtype, cache_only, tries)
+end
+
+
+function _M.toip(name, port, cache_only, tries)
+  return dns_client:resolve_address(name, port, cache_only, tries)
+end
+
+
+-- "_ldap._tcp.example.com:33" -> "_ldap._tcp.example.com:SRV"
+local function format_key(key)
+  local qname, qtype = key:match("([^:]+):(%-?%d+)")  -- match "(qname):(qtype)"
+  return qtype and qname .. ":" .. (TYPE_TO_NAME[tonumber(qtype)] or qtype)
```

Comment:
```suggestion
  return qtype and qname .. "|" .. (TYPE_TO_NAME[tonumber(qtype)] or qtype)
```

Not a big fan of the `:` separator as IPv6 address could also contain `:`, we should use a different separator (that can not appear inside valid domain names).

Reply:
resolved, also fixed bug of `key:match("([^:]+):(%-?%d+)")` in case of IPv6 qname.